app/src/main/java/com/oracle/javafx/scenebuilder/app/util/TextInputControlPaster.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.oracle.javafx.scenebuilder.app.util;
+
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import javafx.scene.control.IndexRange;
+import javafx.scene.control.TextInputControl;
+import javafx.scene.input.Clipboard;
+
+public class TextInputControlPaster {
+    private final TextInputControl tic;
+    private final Supplier<String> textSource;
+    
+    /**
+     * This class provides a workaround for the issue in JavaFX
+     * which prevents Scene Builder from pasting text values properly
+     * in TextInputControl elements on MacOS. If attempted, clip board
+     * contents is always pasted twice.
+     * 
+     * @see https://bugs.openjdk.java.net/browse/JDK-8205915
+     */
+    public TextInputControlPaster(TextInputControl tic) {
+        this(tic, ()->Clipboard.getSystemClipboard().toString());
+    }
+    
+    TextInputControlPaster(TextInputControl tic, Supplier<String> textSource) {
+        this.tic = Objects.requireNonNull(tic);
+        this.textSource = Objects.requireNonNull(textSource);
+    }
+    
+    public void pasteConditionally(boolean condition) {
+        if (condition) {
+            String text = textSource.get();
+            pasteWorkaroundOnMacOs(text);
+        } else {
+            tic.paste();
+        }
+    }
+
+    protected void pasteWorkaroundOnMacOs(String text) {
+        if (text != null) {
+            if (tic.getSelection().getLength() > 0) {
+                replaceSelection(text);
+            } else {                
+                insertAtCaret(text);
+            }
+        }
+    }
+
+    protected void insertAtCaret(String text) {
+        int caret = tic.getCaretPosition();
+        insertText(text, caret, caret);
+    }
+
+    protected void replaceSelection(String text) {
+        IndexRange indexRange = tic.getSelection();
+        insertText(text, indexRange.getStart(), indexRange.getEnd());
+    }
+    
+    private void insertText(String text, int start, int end) {
+        String ticText = tic.getText();
+        String prefix = ticText.substring(0, start);
+        String suffix = ticText.substring(end, ticText.length());
```

Comment:
`ticText.length()` is not required?

Reply:
This code has been removed, it is not needed anymore.